src/Client/V2/Access/DatabaseAccess.cs

```diff
@@ -389,11 +411,211 @@ public virtual DatabaseQueryResult Query( string databaseId, DatabaseQuery query
+        /// Queries the database for information, composing the query with information provided in the
+        /// specified query structure. This method returns an async enumerable where each item is a
+        /// read only dictionary that represents a single row. The keys in the returned dictionary
+        /// are the field ids, and the values are the values for each field represented by the field id.
+        /// This will throw an <see cref="System.ArgumentException"/> if a field is specified more than
+        /// once in the query due to attempting to add duplicate dictionary keys and QueryValuesAsync
+        /// should be used instead. This is a convenience wrapper that calls both StartQueryAsync
+        /// and StopQueryAsync after completing the read, but if you need access to the <see cref="AsyncQueryInfo"/>
+        /// those methods need to be called manually with <see cref="ReadQueryDictionaryAsync(string, AsyncQueryInfo, int, CallContext)"/>
+        /// to read the data.
+        /// </summary>
+        /// <param name="databaseId">
+        /// The unique identifier of the EMS database to query.
+        /// </param>
+        /// <param name="query">
+        /// The information used to construct a query for which results are returned.
+        /// </param>
+        /// <param name="rowsPerCall">
+        /// The number of rows to read for each API call. Setting this to a lower value will return
+        /// async enumerable values more quickly but may be slower overall due to requiring more API
+        /// requests due to smaller pages.
+        /// </param>
+        /// <param name="context">
+        /// The optional call context to include.
+        /// </param>
+        public virtual async IAsyncEnumerable<IReadOnlyDictionary<string, object>> QueryDictionaryAsync( string databaseId, DatabaseQuery query, int rowsPerCall = DefaultRowsPerPage, CallContext context = null )
+        {
+            AsyncQueryInfo info = null;
+            try
+            {
+                info = await StartQueryAsync( databaseId, query, context );
+                string[] orderedIds = GetOrderedFieldIds( info );
+
+                int startingRow = 0;
+                bool moreToRead = true;
+                while( moreToRead )
+                {
+                    AsyncQueryData data = await ReadQueryAsync( databaseId, info.Id, startingRow, startingRow + rowsPerCall - 1, context );
+                    foreach( ICollection<object> row in data.Rows )
+                    {
+                        int i = 0;
+                        var result = new Dictionary<string, object>();
+                        foreach( object value in row )
+                        {
+                            result.Add( orderedIds[i], ScrubDatabaseValue( value ) );
+                            i++;
+                        }
+
+                        yield return result;
+                    }
+                    startingRow += rowsPerCall;
```

Comment:
I know this is how the API works, but maybe adding in the number of rows returned vs. the rowsPerCall might be a bit safer? Safer in terms of if the API behavior changes.

Probably this is a stupid comment, because if we changed that other shit would break. The only time it returns less right now is if there are no more rows.

Reply:
I thought about this some and am inclined to leave it as-is, mostly because there are so many other pieces of code out there that look like this I'd be afraid to change the API at all (like you said)